### ISSUE TITLE
object-curly-newline: multiline should be aware of nesting

### ISSUE THREAD

**Issue by balupton** [eslint/eslint#7434]
**Tell us about your environment**
- **ESLint Version:** $ v3.8.1
- **Node Version:** v6.9.1
- **npm Version:** 3.10.9

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:** https://github.com/bevry/base/blob/bfbac29c5342d1c379f4a587a72172d8f95ee3b0/.eslintrc.js#L642-L643

```
        // Object indentation should be consistent within the object
        'object-curly-newline': [2, {multiline: true}],
```

**What did you do? Please include the actual source code causing the issue.**

``` js
const writetree = {
    'a file': 'content of a file',
    'a directory': {
        'a sub file of a directory': 'content of a sub file of a directory',
        'another sub file of a directory': 'content of another sub file of a directory'
    },
    '.a hidden directory': {
        'a sub file of a hidden directory': 'content of a sub file of a hidden directory'
    },
    'a specific ignored file': 'content of a specific ignored file'
}
```

**What did you expect to happen?**

It to remain the same, instead it got converted with `--fix` to:

``` javascript
const writetree = {
    'a file': 'content of a file',
    'a directory': {
        'a sub file of a directory': 'content of a sub file of a directory',
        'another sub file of a directory': 'content of another sub file of a directory'
    },
    '.a hidden directory': {'a sub file of a hidden directory': 'content of a sub file of a hidden directory'},
    'a specific ignored file': 'content of a specific ignored file'
}
```

**What actually happened? Please include the actual, raw output from ESLint.**

This is the failed linting with the desired indentation, not the undesired fixed indentation:

```
$ eslint source/

/Users/balupton/Projects/active/watchr/source/test.js
  34:25  error  Unexpected line break after this opening brace   object-curly-newline
  36:2   error  Unexpected line break before this closing brace  object-curly-newline

✖ 2 problems (2 errors, 0 warnings)

```

**Comment by not-an-aardvark** [comment]
Thanks for the report. However, I'm confused about why this would be a bug; it seems like it's working as intended.

Based on the [docs](http://eslint.org/docs/rules/object-curly-newline#multiline), looks like your `a hidden directory` object incorrectly has newlines after the curly brackets, so the fixer fixes that for you.

**Comment by balupton** [comment]
Thanks. This isn't really a bug per se, but perhaps a change request for more desirable behaviour. I can't find anyway to make my desired result an enforced thing - and I thought the multiline option should be this enforcement - with the interpretation, of an objects scope being the entire object tree, not just the current object.

**Comment by not-an-aardvark** [comment]
Does `/*eslint object-curly-newline: ["error", "always"]*/` yield your desired result? (See [here](http://eslint.org/docs/rules/object-curly-newline#always) for examples with that option.)

**Comment by balupton** [comment]
Well, that would change `{a:1}` to:

```
{
  a: 1
}
```

which isn't desired.

All in all, this is the desire:

```
// Correct
o = {a: 1}
o = {a: 1, b:2}
o = {
  a: 1,
  b: 2
}
o = {a: 1, b: {c: 1}}
o = {
 a: 1,
 b: {
   c: 1
 }
}
o = {a: 1, b: {c: 1, d: 2}}
o = {
 a: 1,
 b: {
   c: 1,
   d: 2
 }
}

// Incorrect
o = {
  a: 1,
  b: {c: 1}
}
o = {
  a: 1,
  b: {c: 1, d: 2}
}
```

So either an object tree should all be singleline, or all be multiline

**Comment by kaicataldo** [comment]
Closing this as discussion has stalled and it doesn't appear there is enough interest, either from the team or from the community, to continue the discussion at this time. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. Please feel free to open a new issue with a proposal to continue this discussion if this is something you would like to continue advocating for.
